### PR TITLE
Add option to run TensorFlow job on the preferred device (via Scope)

### PR DESCRIPTION
**Pull request by zaleslaw**
Related to the issue #77

### CODE REVIEW

File: tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java [hunk 1]
```diff
@@ -19,11 +19,8 @@
 
 import java.nio.charset.Charset;
 import java.util.List;
-import org.tensorflow.DataType;
-import org.tensorflow.EagerSession;
-import org.tensorflow.ExecutionEnvironment;
-import org.tensorflow.Operand;
-import org.tensorflow.Tensor;
+
+import org.tensorflow.*;
```
Comment (karllessard) [inline tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java:23]: Let's avoid wildcard imports, as per [Google Java style guide](https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports) suggestion.

File: tensorflow-core/tensorflow-core-api/src/gen/java/org/tensorflow/op/math/Abs.java [hunk 2]
```diff
@@ -50,6 +50,7 @@
   @Endpoint(describeByClass = true)
   public static <T extends TNumber> Abs<T> create(Scope scope, Operand<T> x) {
     OperationBuilder opBuilder = scope.env().opBuilder("Abs", scope.makeOpName("Abs"));
+    opBuilder.setDevice(scope.makeDeviceString());
```
Comment (karllessard) [inline tensorflow-core/tensorflow-core-api/src/gen/java/org/tensorflow/op/math/Abs.java:53]: All generated ops should have been update by your change in the op generator above, they should be all updated by this PR as well, not only this one (there must be 1000+ files to be added).

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/Scope.java [hunk 3]
```diff
@@ -121,7 +123,12 @@ public Scope withSubScope(String childScopeName) {
    * @throws IllegalArgumentException if the name is invalid
    */
   public Scope withName(String opName) {
-    return new Scope(env, nameScope.withName(opName), controlDependencies);
+    return new Scope(env, nameScope.withName(opName), controlDependencies, deviceSpec);
+  }
+
+  // TODO: add docs
```
Comment (karllessard) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/Scope.java:129]: TODO: add docs :)

File: tensorflow-core/tensorflow-core-api/src/test/java/org/tensorflow/op/core/ConstantTest.java [hunk 4]
```diff
@@ -23,11 +23,11 @@
...
 import java.nio.IntBuffer;
 import java.nio.LongBuffer;
+import java.util.Collection;
+
 import org.junit.jupiter.api.Test;
-import org.tensorflow.AutoCloseableList;
-import org.tensorflow.Graph;
-import org.tensorflow.Session;
-import org.tensorflow.Tensor;
+import org.tensorflow.*;
```
Comment (karllessard) [inline tensorflow-core/tensorflow-core-api/src/test/java/org/tensorflow/op/core/ConstantTest.java:29]: No wildcard imports

File: tensorflow-core/tensorflow-core-api/src/test/java/org/tensorflow/op/core/ConstantTest.java [hunk 5]
```diff
@@ -70,6 +71,46 @@ public void createInts() {
...
+              .build()
+              .output(0);
+      DeviceSpec deviceSpec = DeviceSpec.newBuilder()
+              .job("localhost")
+              .replica(0)
+              .task(1)
+              .deviceType(DeviceSpec.DeviceType.CPU)
+              .build();
+
+      //DeviceSpec deviceSpec = DeviceSpec.newBuilder().build();
```
Comment (karllessard) [inline tensorflow-core/tensorflow-core-api/src/test/java/org/tensorflow/op/core/ConstantTest.java:101]: This comment is probably a leftover

File: tensorflow-core/tensorflow-core-generator/src/main/java/org/tensorflow/processor/operator/OperatorProcessor.java [hunk 6]
```diff
@@ -537,6 +538,18 @@ private static TypeSpec buildTopClass(OpsSpec spec) {
...
+                    .addModifiers(Modifier.PUBLIC)
+                    .addParameter(T_DEVICE_SPEC, "deviceSpec")
+                    .returns(T_OPS)
+                    .addStatement("return new Ops(scope.withDevice(deviceSpec))")
+                    .addJavadoc(
+                            "Returns an API that uses the provided DeviceSpec for an op.\n\n"
+                                    + "@see {@link $T#withDevice(DeviceSpec)}\n",
+                            T_SCOPE)
+                    .build());
+
```
Comment (karllessard) [inline tensorflow-core/tensorflow-core-generator/src/main/java/org/tensorflow/processor/operator/OperatorProcessor.java:552]: The `Ops` class should now have this new method in it but it does not appear in the version you've committed.

Also, I would rephrase its documentation for something like this:

"Returns an API that places the created operations on the device(s) matching the provided spec. @see..."

**Review by karllessard — CHANGES_REQUESTED** [review]

**Comment by zaleslaw** [comment]
Thank you, it was a draft PR with code changes, but without code
generation, I will update the PR and notify you

вс, 29 нояб. 2020 г., 1:13 Karl Lessard <notifications@github.com>:

**Comment by karllessard** [comment]
Ah, did you marked it as a draft when you've created your PR? My GitHub iPad app did not tell me this, maybe a missing feature in the app. But yes, I think with only these small changes, your PR will be ready to be merged, thanks!

**Comment by zaleslaw** [comment]
No, I didn't mark, sorry for that, but your comments are useful, because
now it's clear what should be checked for successful PR

вс, 29 нояб. 2020 г., 17:35 Karl Lessard <notifications@github.com>:

**Comment by zaleslaw** [comment]
@karllessard I've updated the PR and fixed all your proposals, please have a look

File: tensorflow-core/tensorflow-core-api/src/bazel/op_generator/op_generator.cc [hunk 7]
```diff
@@ -246,6 +246,10 @@ void RenderFactoryMethods(const OpSpec& op, const Type& op_class,
       writer->EndLine();
     }
   }
+  // Add Device String
+  writer->Append("opBuilder.setDevice(scope.makeDeviceString());");
+  writer->EndLine();
+
```
Comment (karllessard) [inline tensorflow-core/tensorflow-core-api/src/bazel/op_generator/op_generator.cc:252]: Hey @zaleslaw , I have another change to propose you that I just thought of. I think if we merge together `scope.applyControlDependencies` and this new step for adding the device string to the `opBuilder` would make the whole process of building an op simpler, especially with non-generated ones.

So what about that:
1. We rename `scope.applyControlDependencies(opBuilder)` to `scope.apply(opBuilder)`
2. In `Scope.apply`, we apply both control dependencies, the device string (and eventually more data if needed)

By doing so, we also won't need `scope.makeDeviceString()` anymore. Let me know what you think

**Comment by zaleslaw** [comment]
Need a time to think about it



пн, 30 нояб. 2020 г. в 17:19, Karl Lessard <notifications@github.com>:

**Comment by zaleslaw** [comment]
@karllessard I've updated a PR according your refactoring idea, but I have some troubles with quick build
"Failed to execute goal on project tensorflow-core-api: Could not resolve dependencies for project org.tensorflow:tensorflow-core-api:jar:0.3.0-SNAPSHOT: Could not find artifact org.tensorflow:tensorflow-core-api:jar:linux-x86_64:0.3.0-SNAPSHOT in ossrh-snapshots (https://oss.sonatype.org/content/repositories/snapshots) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal on project tensorflow-core-api: Could not resolve dependencies for project org.tensorflow:tensorflow-core-api:jar:0.3.0-SNAPSHOT: Could not find artifact org.tensorflow:tensorflow-core-api:jar:linux-x86_64:0.3.0-SNAPSHOT in ossrh-snapshots (https://oss.sonatype.org/content/repositories/snapshots)"

Could you pleas help with this issue?

**Comment by karllessard** [comment]
@zaleslaw since your PR changes the generated files, we shouldn't use quick build anyway, I'll force a full build and see how it goes (independently to this, I suspect the missing artifact is due again to our trouble building all native artifacts on GitHub actions...)

**Review by karllessard — APPROVED** [review]

**Comment by karllessard** [comment]
Full build passed (well partially, GPU-MKL timed out again on Linux and it's unrelated), thanks @zaleslaw !

**Comment by zaleslaw** [comment]
Happy to hear!

**Comment by aday00** [comment]
Thanks @zaleslaw and @karllessard!!
